src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -904,6 +904,28 @@ private static boolean doesExtendFXApplication(Class<?> mainClass) {
         return false;
     }
 
+    /*
+     * main type flags
+     */
+    private static final int MAIN_WITHOUT_ARGS = 1;
+    private static final int MAIN_NONSTATIC = 2;
+    private static int mainType = 0;
```

Comment:
Thinking about this some more, would it not be possible to just use the mainMethod directly down in C?

Reply:
The changes JEP 463 went through many iterations, it was a fine balance of avoiding too many transitions and upscales, and at the same time, have something that can be maintained. There are several JBS issues on this issue now, probably because there just wasn't enough tests introduced with the JEP.

Reply:
FWIW, just ran into this as well. I was trying to do `System::loadLibrary` and had no clue why I could load non-existent libraries. Turns out I couldn't :-) but the error was never reported back to me. Confusing.